pliers/extractors/audio.py

```diff
@@ -488,4 +494,88 @@ class PercussiveExtractor(LibrosaFeatureExtractor):
     For details on argument specification visit:
     https://librosa.github.io/librosa/effect.html.'''
 
+
     _feature = 'percussive'
+
+
+class AudiosetLabelExtractor(AudioExtractor):
+
+    ''' Extract probability of audio event classes based on AudioSet-YouTube 
+    corpus using a YAMNet architecture. Code available at:
+    https://github.com/tensorflow/models/tree/master/research/audioset/yamnet 
+
+    Args:
+    hop_size (float): size of the window (in seconds) on which label extraction
+        is performed.
+    top_n (int): specifies how many of the highest label probabilities are 
+        returned. If not defined, returns probabilities for all labels.
+    spectrogram (bool): if True, plots mel-spectrogram of the audio stimulus.
+    yamnet_kwargs (optional): Optional named arguments that modify input 
+        parameters for the model (see params.py file in yamnet repository)
+    '''
+
+    _log_attributes = ('params', 'top_n')
+
+    def __init__(self, hop_size=0.1, top_n=None, 
+                 spectrogram=False, **yamnet_kwargs):
+        verify_dependencies(['yamnet'])
+        verify_dependencies(['tensorflow'])
+
+        MODULE_PATH = path.dirname(yamnet.__file__)
+        WEIGHTS_PATH = path.join(MODULE_PATH, 'yamnet.h5')
+        LABELS_PATH = path.join(MODULE_PATH, 'yamnet_class_map.csv')
+
+        self.spectrogram = spectrogram
+        self.params = yamnet.params
+        self.params.PATCH_HOP_SECONDS = hop_size
+
+        if yamnet_kwargs:
+            for par in self.params.__dict__:
+                if yamnet_kwargs[par]:
+                    setattr(self.params, par, yamnet_kwargs[par])
+
+        self.tf_graph = tf.Graph()
+        with self.tf_graph.as_default():
+            self.model = yamnet.yamnet_frames_model(self.params)
+            self.model.load_weights(WEIGHTS_PATH)
+
+        self.labels = pd.read_csv(LABELS_PATH)['display_name'].tolist()
+        self.params = {par: val for par,val in self.params.__dict__.items()
+                       if par.isupper()}
+        self.top_n = top_n if top_n else len(self.labels)
+        super(AudiosetLabelExtractor, self).__init__()
+
+    def _extract(self, stim):
+        
+        data = stim.data
+        self.params['SAMPLE_RATE'] = stim.sampling_rate
```

Comment:
Are you sure its OK to do this? You explained to me that it was only change how the spectrogram is created, so if you're positive, sounds good to me. I'm just noting this comment in params.py:


> The following hyperparameters (except PATCH_HOP_SECONDS) were used to train YAMNet,
> so expect some variability in performance if you change these. The patch hop can
> be changed arbitrarily: a smaller hop should give you more patches from the same
> clip and possibly better performance at a larger computational cost.

Reply:
Might be worth testing the same audio clip with two sampling rates

Reply:
> Oh, and this works in TF 1 & 2 right?

yes, found a workaround to `yamnet` versioning issue by explicitly calling tf graph. With will keep ensuring that it's compatible with both. Also raised the issue on yamnet repo (https://github.com/tensorflow/models/issues/8157), something might happen on that front too.

Reply:
will make sure to test it thoroughly (and look into an issue with output values I just noticed) and call another round of review :)

Reply:
okay, I've looked into this and I think I'm on top of it.
Tl;dr -> it's okay to have other `SAMPLE_RATE` than 16000Hz, as a) size of input patches will still be the same; b) results should not be significantly affected. Caveat is that `SAMPLE_RATE` has to be twice as large as `MEL_MAX_HZ`. Writing down a lengthy explanation below, mostly as a reminder for myself. :)

### (Why) the model is compatible with different sampling rates
Yamnet is compatible with different sampling rates, although it's been trained with audio sampled at 16000Hz. 

The input sampling rate is only relevant in the preprocessing step, where the waveform is first passed through STFT to extract spectrogram, which is then converted into a mel-scale spectrogram. 

The sampling rate parameter is used to compute how many samples in the stimulus go into a STFT window, but the size of the window is independent from `SAMPLE_RATE` and explicitly encoded in a separate parameter. Therefore, different sampling rates do not affect how the waveform is binned in the **temporal** domain. 

What `SAMPLE_RATE` does implicitly influence, though, is the (number, values and resolution of) **frequency bins** used for the Fourier transform. Higher `SAMPLE_RATE` correspond to higher frequency resolution in the spectrogram. 

The second pass is transforming the spectrogram into a mel-scale spectrogram. The trick here is that, regardless of how many and which frequency bins you have in the spectrogram, the number of frequency bands in the mel spectrogram is fixed by the parameters `MEL_BANDS` (= number of mel-frequency bins, set to 64 for training) and its range is determined by `MEL_MIN_HZ` / `MEL_MAX_HZ` (set to 125/7500 Hz for training). Input spectrograms of size [n_freq_bins, n_time_bins] will always output [n_mel_bands, n_time_bins], where n_mel_bands is independent of n_freq_bins. That's why different sampling rates would yield still fit model requirements in terms of input size. 

### Different sampling rates should have no impact on the results
Although the input shape is not affected, having different sampling rates on the same waveform _might_ still yield slightly differences in the spectrograms (as FFT frequency resolution will differ).
But as long as the mel frequency range (`MEL_MIN_HZ` and `MEL_MAX_HZ`) is kept close to that used for training these differences shouldn't have any impact. 

### Which of the params can be changed?
- `MEL_BANDS` has to be 64 as the model expects this in its input
- `MEL_MIN_HZ` and `MEL_MAX_HZ` can be changed, with the caveat that model performance can actually be affected and that `MEL_MAX_HZ` has to be less than half of the sampling rate (Nyquist freq).
- `SAMPLE_RATE` can be modified and adapted to actual sampling rate, with the constraint that it has to be at least twice as large as` MEL_MAX_HZ`.

### Re: pliers extractor
- Would probably add a warning if `SAMPLE_RATE` ≠ 16000Hz telling the user that the model was trained on 1600Hz, different sampling rates may have a minimal impact on the results and it could be an idea to resample;
- Would also add a pliers error if `SAMPLE_RATE` < 2 * `MEL_MAX_HZ`, warning that it may be a bad idea to have lower sampling rate and suggesting to either upsample or change `MEL_MAX_HZ` to a suitable value, although this may reduce accuracy of the predictions